source/kernel/src/trace.rs

```diff
@@ -82,29 +99,116 @@ impl SerialCollector {
             }
         };
         wgr.commit(len);
+        self.in_send.store(false, Ordering::Release);
 
         // return true if we committed a non-zero number of bytes.
         len > 0
     }
 
-    async fn worker(rx: bbq::Consumer, port: serial_mux::PortHandle) {
-        loop {
-            let rgr = rx.read_grant().await;
+    async fn worker(
+        &'static self,
+        rx: bbq::Consumer,
+        port: serial_mux::PortHandle,
+        k: &'static crate::Kernel,
+    ) {
+        use futures::FutureExt;
+        use maitake::time;
+        use postcard::accumulator::{CobsAccumulator, FeedResult};
+
+        // we probably won't use 256 whole bytes of cobs yet since all the host
+        // -> target messages are quite small
+        let mut cobs_buf: CobsAccumulator<16> = CobsAccumulator::new();
+        let mut read_level = |rgr: bbq::GrantR| {
+            let mut window = &rgr[..];
             let len = rgr.len();
-            port.send(&rgr[..]).await;
+            'cobs: while !window.is_empty() {
+                window = match cobs_buf.feed_ref::<HostRequest>(window) {
+                    FeedResult::Consumed => break 'cobs,
+                    FeedResult::OverFull(new_wind) => new_wind,
+                    FeedResult::DeserError(new_wind) => new_wind,
+                    FeedResult::Success { data, remaining } => {
+                        match data {
+                            HostRequest::SetMaxLevel(lvl) => {
+                                let level = lvl.map(|lvl| lvl as u8).unwrap_or(5);
```

Comment:
Maybe note what "5" is here?

Reply:
changed this to actually use the enum variant